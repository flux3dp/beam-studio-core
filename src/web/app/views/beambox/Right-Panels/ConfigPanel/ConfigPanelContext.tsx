import { createContext, Dispatch } from 'react';

import { DataType, dataKey, defaultConfig } from 'helpers/layer/layer-config-helper';
import { ILayerConfig } from 'interfaces/ILayerConfig';

interface State extends ILayerConfig {
  selectedItem?: string;
}

export const getDefaultState = (): State => {
  const keys = Object.keys(defaultConfig);
  const initState = {} as State;
  keys.forEach((key) => {
    // Handle DataType and state type mismatch
    initState[dataKey[key]] = { value: defaultConfig[key as DataType] };
  });
  return initState;
};

export type Action = {
  type: 'update';
  payload: State;
} | {
  type: 'change';
  payload: {
    selectedItem?: string;
  } & {
    [key in keyof ILayerConfig]?: number | string;
  };
} | {
  type: 'rename';
  payload: string;
};

export const reducer = (state: State, action: Action): State => {
  if (action.type === 'update') return { ...state, ...action.payload };
  if (action.type === 'change') {
    const { payload } = action;
    const newState = { ...state };
    Object.keys(payload).forEach((key) => {
      if (key !== 'selectedItem') newState[key] = { value: payload[key] };
      else newState[key] = payload[key];
    });
    return newState;
  }
  if (action.type === 'rename') {
    const { payload } = action;
    const newState = { ...state };
    newState.configName = { value: payload };
    newState.selectedItem = payload;
    return newState;
  }
  return state;
};

interface Context {
  state: State;
  selectedLayers: string[];
  dispatch: Dispatch<Action>;
  simpleMode?: boolean;
  initState: (layers?: string[]) => void;
}

export default createContext<Context>({} as Context);
