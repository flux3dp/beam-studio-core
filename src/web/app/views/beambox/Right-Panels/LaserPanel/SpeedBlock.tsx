import classNames from 'classnames';
import React, { useMemo } from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import doLayersContainsVector from 'helpers/layer/check-vector';
import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';
import storage from 'implementations/storage';

import styles from './SpeedBlock.module.scss';

interface Props {
  layerNames: string[];
  speed: number;
  hasMultipleValue: boolean;
  onChange: (val: number) => void;
}

const MIN_VALUE = 1;

const SpeedBlock = ({ layerNames, speed, hasMultipleValue, onChange }: Props): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;

  const { display: displayUnit, decimal } = useMemo(() => {
    const unit: 'mm' | 'inches' = storage.get('default-units') || 'mm';
    const display = { mm: 'mm/s', inches: 'in/s' }[unit];
    const d = { mm: 1, inches: 2 }[unit];
    return { display, decimal: d };
  }, []);
  const maxValue = BeamboxPreference.read('workarea') === 'fhexa1' ? 900 : 300;

  const hasVector = doLayersContainsVector(layerNames);

  return (
    <div className={styles.panel}>
      <span className={styles.title}>{t.speed}</span>
      <UnitInput
        id="speed"
        className={{ [styles.input]: true }}
        min={MIN_VALUE}
        max={maxValue}
        unit={displayUnit}
        defaultValue={speed}
        getValue={onChange}
        decimal={decimal}
        displayMultiValue={hasMultipleValue}
      />
      <input
        id="speed_value"
        className={classNames({ [styles['speed-for-vector']]: hasVector })}
        type="range"
        min={MIN_VALUE}
        max={maxValue}
        step={1}
        value={speed}
        onChange={(e) => onChange(Number(e.target.value))}
      />
      {hasVector && speed > 20 && BeamboxPreference.read('vector_speed_contraint') !== false ? (
        <div className={styles.warning}>
          <div className={styles['warning-icon']}>!</div>
          <div className={styles['warning-text']}>{t.speed_contrain_warning}</div>
        </div>
      ) : null}
    </div>
  );
};

export default SpeedBlock;
