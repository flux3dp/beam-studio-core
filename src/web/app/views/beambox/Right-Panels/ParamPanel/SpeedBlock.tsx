import classNames from 'classnames';
import React, { useMemo } from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import constant from 'app/actions/beambox/constant';
import doLayersContainsVector from 'helpers/layer/check-vector';
import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';
import storage from 'implementations/storage';
import { IConfig } from 'interfaces/ILayerConfig';

import styles from './Block.module.scss';

interface Props {
  layerNames: string[];
  speed: IConfig<number>;
  onChange: (val: number) => void;
}

const SpeedBlock = ({ layerNames, speed, onChange }: Props): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;
  const { value, hasMultiValue } = speed;

  const { display: displayUnit, decimal } = useMemo(() => {
    const unit: 'mm' | 'inches' = storage.get('default-units') || 'mm';
    const display = { mm: 'mm/s', inches: 'in/s' }[unit];
    const d = { mm: 1, inches: 2 }[unit];
    return { display, decimal: d };
  }, []);
  const model = BeamboxPreference.read('workarea');
  const enableLowSpeed = BeamboxPreference.read('enable-low-speed');
  const maxValue = constant.dimension.getMaxSpeed(model);
  const minValue = enableLowSpeed ? 1 : 3;
  const hasVector = doLayersContainsVector(layerNames);
  let warningText = '';
  if (hasVector && value > 20 && BeamboxPreference.read('vector_speed_contraint') !== false) {
    warningText = t.speed_contrain_warning;
  } else if (value < 3 && enableLowSpeed) {
    warningText = t.low_speed_warning;
  }

  return (
    <div className={styles.panel}>
      <span className={styles.title}>{t.speed}</span>
      <UnitInput
        id="speed"
        className={{ [styles.input]: true }}
        min={minValue}
        max={maxValue}
        unit={displayUnit}
        defaultValue={value}
        getValue={onChange}
        decimal={decimal}
        displayMultiValue={hasMultiValue}
        step={10 ** -decimal}
      />
      <input
        id="speed_value"
        className={classNames({ [styles['speed-for-vector']]: hasVector })}
        type="range"
        min={minValue}
        max={maxValue}
        step={0.1}
        value={value}
        onChange={(e) => onChange(Number(e.target.value))}
      />
      {warningText ? (
        <div className={styles.warning}>
          <div className={styles['warning-icon']}>!</div>
          <div className={styles['warning-text']}>{warningText}</div>
        </div>
      ) : null}
    </div>
  );
};

export default SpeedBlock;
