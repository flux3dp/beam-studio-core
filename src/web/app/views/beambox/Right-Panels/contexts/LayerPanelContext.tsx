import * as React from 'react';
import { ILayerPanelContext } from 'interfaces/IContext';

export const LayerPanelContext = React.createContext<ILayerPanelContext>(null);

interface Props {
  children: React.ReactNode;
}

export class LayerPanelContextProvider extends React.Component<Props> {
  private contextValue = null;

  private selectedLayers: string[] = [];

  constructor(props) {
    super(props);
    this.updateContextValue();
  }

  updateContextValue = (): void => {
    const {
      selectedLayers,
      setSelectedLayers,
      updateLayerPanel,
    } = this;
    this.contextValue = {
      selectedLayers,
      setSelectedLayers,
      updateLayerPanel,
    };
  };

  setSelectedLayers = (selectedLayers: string[]): null => {
    this.selectedLayers = [...selectedLayers];
    this.updateContextValue();
    this.forceUpdate();
    return null;
  };

  updateLayerPanel = (): void => {
    this.forceUpdate();
  };

  render() {
    const { children } = this.props;
    return (
      <LayerPanelContext.Provider value={this.contextValue}>
        {children}
      </LayerPanelContext.Provider>
    );
  }
}
