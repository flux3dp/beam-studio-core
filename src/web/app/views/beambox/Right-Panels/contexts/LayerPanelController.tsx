import eventEmitterFactory from 'helpers/eventEmitterFactory';

const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

const updateLayerPanel = (): void => {
  layerPanelEventEmitter.emit('UPDATE_LAYER_PANEL');
};

const getSelectedLayers = (): string[] => {
  const response = {
    selectedLayers: [],
  };
  layerPanelEventEmitter.emit('GET_SELECTED_LAYERS', response);
  return response.selectedLayers;
};

const setSelectedLayers = (selectedLayers: string[]): void => {
  layerPanelEventEmitter.emit('SET_SELECTED_LAYERS', selectedLayers);
};

export default {
  updateLayerPanel,
  getSelectedLayers,
  setSelectedLayers,
};
