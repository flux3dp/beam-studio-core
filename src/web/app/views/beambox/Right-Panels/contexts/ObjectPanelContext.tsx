import React from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';

interface IObjectPanelContext {
  polygonSides: number;
  dimensionValues: any;
  updateDimensionValues: (newValues: any) => void;
  getDimensionValues: (response: {
    dimensionValues: any,
  }, key?: string,) => any;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  polygonSides: 5,
  dimensionValues: {},
  updateDimensionValues: () => { },
  getDimensionValues: () => { },
  updateObjectPanel: () => { },
});
const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface State {
  lastUpdateTime: number;
  polygonSides: number;
}

export class ObjectPanelContextProvider extends React.Component<any, State> {
  private dimensionValues: any;

  private updateTimeout: NodeJS.Timeout;

  constructor(props) {
    super(props);
    this.dimensionValues = {};
    this.state = {
      lastUpdateTime: Date.now(),
      polygonSides: 5,
    };
  }

  componentDidMount() {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', this.updateDimensionValues.bind(this));
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', this.getDimensionValues.bind(this));
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', this.updateObjectPanel.bind(this));
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', this.updatePolygonSides.bind(this));
  }

  componentWillUnmount() {
    objectPanelEventEmitter.removeAllListeners();
  }

  updateDimensionValues = (newValues: any): void => {
    this.dimensionValues = {
      ...this.dimensionValues,
      ...newValues,
    };
  };

  updatePolygonSides = (polygonSides: number): void => {
    this.setState({
      polygonSides,
    });
  };

  getDimensionValues = (response: {
    dimensionValues: any,
  }, key?: string): void => {
    response.dimensionValues = key ? this.dimensionValues[key] : this.dimensionValues;
  };

  updateObjectPanel = (): void => {
    clearTimeout(this.updateTimeout);
    const time = Date.now();
    const { lastUpdateTime } = this.state;
    if (time - lastUpdateTime >= minRenderInterval) {
      this.setState({ lastUpdateTime: time });
    } else {
      this.updateTimeout = setTimeout(() => {
        this.setState({
          lastUpdateTime: lastUpdateTime + minRenderInterval,
        });
      }, lastUpdateTime + minRenderInterval - time);
    }
  };

  render(): JSX.Element {
    const { children } = this.props;
    const { polygonSides } = this.state;
    const {
      dimensionValues,
      updateDimensionValues,
      getDimensionValues,
      updateObjectPanel,
    } = this;
    return (
      <ObjectPanelContext.Provider value={{
        polygonSides,
        dimensionValues,
        updateDimensionValues,
        getDimensionValues,
        updateObjectPanel,
      }}
      >
        {children}
      </ObjectPanelContext.Provider>
    );
  }
}
