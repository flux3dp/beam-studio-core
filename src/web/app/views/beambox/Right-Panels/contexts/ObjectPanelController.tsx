import eventEmitterFactory from 'helpers/eventEmitterFactory';

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');

const updateDimensionValues = (newValue: any): void => {
  objectPanelEventEmitter.emit('UPDATE_DIMENSION_VALUES', newValue);
};

const getDimensionValues = (key: string): any => {
  const response = {
    dimensionValues: {},
  };
  objectPanelEventEmitter.emit('GET_DIMENSION_VALUES', key, response);
  return response.dimensionValues;
};

const updateObjectPanel = (): void => {
  objectPanelEventEmitter.emit('UPDATE_OBJECT_PANEL');
};

export default {
  updateObjectPanel,
  updateDimensionValues,
  getDimensionValues,
};
