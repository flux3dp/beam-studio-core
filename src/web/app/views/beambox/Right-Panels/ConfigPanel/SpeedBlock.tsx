import classNames from 'classnames';
import React, { memo, useCallback, useContext, useMemo } from 'react';
import { Button, Popover } from 'antd-mobile';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import configOptions from 'app/constants/config-options';
import constant from 'app/actions/beambox/constant';
import doLayersContainsVector from 'helpers/layer/check-vector';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import LayerModule from 'app/constants/layer-module/layer-modules';
import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import objectPanelItemStyles from 'app/views/beambox/Right-Panels/ObjectPanelItem.module.scss';
import storage from 'implementations/storage';
import units from 'helpers/units';
import useI18n from 'helpers/useI18n';
import { CUSTOM_PRESET_CONSTANT, DataType, writeData } from 'helpers/layer/layer-config-helper';
import { LayerPanelContext } from 'app/views/beambox/Right-Panels/contexts/LayerPanelContext';
import { ObjectPanelContext } from 'app/views/beambox/Right-Panels/contexts/ObjectPanelContext';

import ConfigPanelContext from './ConfigPanelContext';
import ConfigSlider from './ConfigSlider';
import ConfigValueDisplay from './ConfigValueDisplay';
import styles from './Block.module.scss';

const SpeedBlock = ({
  type = 'default',
}: {
  type?: 'default' | 'panel-item' | 'modal';
}): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;
  const { selectedLayers, state, dispatch, simpleMode = true } = useContext(ConfigPanelContext);
  const { activeKey } = useContext(ObjectPanelContext);
  const visible = activeKey === 'speed';
  const { hasVector } = useContext(LayerPanelContext);
  const timeEstimationButtonEventEmitter = useMemo(
    () => eventEmitterFactory.createEventEmitter('time-estimation-button'),
    []
  );

  const { value, hasMultiValue } = state.speed;
  const module = state.module.value;

  const {
    display: displayUnit,
    decimal,
    calculateUnit: fakeUnit,
  } = useMemo(() => {
    const unit: 'mm' | 'inches' = storage.get('default-units') || 'mm';
    const display = { mm: 'mm/s', inches: 'in/s' }[unit];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const calculateUnit: 'mm' | 'inch' = { mm: 'mm', inches: 'inch' }[unit] as any;
    const d = { mm: 1, inches: 2 }[unit];
    return { display, decimal: d, calculateUnit };
  }, []);
  const workarea = BeamboxPreference.read('workarea');
  const maxValue = useMemo(() => constant.dimension.getMaxSpeed(workarea), [workarea]);
  const workareaMinSpeed = constant.dimension.getMinSpeed(workarea);
  let minValue = workareaMinSpeed;
  const enableLowSpeed = BeamboxPreference.read('enable-low-speed');
  if (minValue > 1 && enableLowSpeed) minValue = 1;
  let warningText = '';
  if (hasVector && value > 20 && BeamboxPreference.read('vector_speed_contraint') !== false) {
    warningText = t.speed_contrain_warning;
  } else if (value < workareaMinSpeed && enableLowSpeed) {
    warningText = t.low_speed_warning;
  }

  const handleChange = (val: number) => {
    dispatch({
      type: 'change',
      payload: { speed: val, configName: CUSTOM_PRESET_CONSTANT },
    });
    timeEstimationButtonEventEmitter.emit('SET_ESTIMATED_TIME', null);
    if (type !== 'modal')
      selectedLayers.forEach((layerName) => {
        writeData(layerName, DataType.speed, val, true);
        writeData(layerName, DataType.configName, CUSTOM_PRESET_CONSTANT);
      });
  };

  const sliderOptions = useMemo(
    () =>
      simpleMode && module === LayerModule.PRINTER
        ? configOptions.getPrintingSpeedOptions(lang)
        : null,
    [simpleMode, module, lang]
  );

  const content = (
    <div className={classNames(styles.panel, styles[type])}>
      <span className={styles.title}>{t.speed}</span>
      <ConfigValueDisplay
        inputId="speed-input"
        type={type}
        max={maxValue}
        min={minValue}
        value={value}
        unit={displayUnit}
        hasMultiValue={hasMultiValue}
        decimal={decimal}
        onChange={handleChange}
        options={sliderOptions}
      />
      <ConfigSlider
        id="speed"
        value={value}
        onChange={handleChange}
        min={minValue}
        max={maxValue}
        step={0.1}
        speedLimit={
          module !== LayerModule.PRINTER &&
          (type === 'modal' ? doLayersContainsVector(selectedLayers) : hasVector)
        }
        options={sliderOptions}
        decimal={decimal}
        unit={displayUnit}
      />
      {warningText ? (
        <div className={styles.warning}>
          <div className={styles['warning-icon']}>!</div>
          <div className={styles['warning-text']}>{warningText}</div>
        </div>
      ) : null}
    </div>
  );

  const getDisplayValue = useCallback(() => {
    const selectedOption = sliderOptions?.find((opt) => opt.value === value);
    if (selectedOption) return selectedOption.label;
    return +units.convertUnit(value, fakeUnit, 'mm').toFixed(decimal);
  }, [decimal, sliderOptions, value, fakeUnit]);

  return type === 'panel-item' ? (
    <Popover visible={visible} content={content}>
      <ObjectPanelItem.Item
        id="speed"
        content={
          <Button
            className={classNames(objectPanelItemStyles['number-item'], styles['display-btn'])}
            shape="rounded"
            size="mini"
            fill="outline"
          >
            <span style={{ whiteSpace: 'nowrap' }}>{getDisplayValue()}</span>
          </Button>
        }
        label={t.speed}
        autoClose={false}
      />
    </Popover>
  ) : (
    content
  );
};

export default memo(SpeedBlock);
