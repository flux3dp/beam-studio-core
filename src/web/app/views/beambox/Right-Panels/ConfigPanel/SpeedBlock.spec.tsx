import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { LayerPanelContext } from 'app/views/beambox/Right-Panels/contexts/LayerPanelContext';

import ConfigPanelContext from './ConfigPanelContext';
import SpeedBlock from './SpeedBlock';

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      laser_panel: {
        speed: 'speed',
        speed_contrain_warning: 'speed_contrain_warning',
      },
    },
  },
}));

jest.mock(
  'app/views/beambox/Right-Panels/ConfigPanel/ConfigSlider',
  () =>
    ({ id, max, min, value, onChange }: any) =>
      (
        <input
          id={id}
          className="mock-config-slider"
          type="range"
          min={min}
          max={max}
          step={0.1}
          value={value}
          onChange={(e) => onChange(Number(e.target.value))}
        />
      )
);

jest.mock(
  'app/widgets/Unit-Input-v2',
  () =>
    ({ id, min, max, unit, defaultValue, decimal, displayMultiValue }: any) =>
      (
        <div>
          MockUnitInput
          <p>id: {id}</p>
          <p>min: {min}</p>
          <p>max: {max}</p>
          <p>unit: {unit}</p>
          <p>defaultValue: {defaultValue}</p>
          <p>decimal: {decimal}</p>
          <p>displayMultiValue: {displayMultiValue}</p>
        </div>
      )
);

jest.mock('app/views/beambox/Right-Panels/ObjectPanelItem', () => ({
  Item: ({ id, content, label, onClick }: any) => (
    <div onClick={onClick}>
      MockObjectPanelItem
      <div>id: {id}</div>
      <div>content: {content}</div>
      <div>label: {label}</div>
    </div>
  ),
}));

const mockWriteData = jest.fn();
jest.mock('helpers/layer/layer-config-helper', () => ({
  CUSTOM_PRESET_CONSTANT: 'CUSTOM_PRESET_CONSTANT',
  DataType: {
    speed: 'speed',
    configName: 'configName',
  },
  writeData: (...args) => mockWriteData(...args),
}));

const mockStorageGet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => mockStorageGet(...args),
}));

const mockPrefRead = jest.fn();
jest.mock('app/actions/beambox/beambox-preference', () => ({
  read: (...args) => mockPrefRead(...args),
}));

const mockSelectedLayers = ['layer1', 'layer2'];
const mockContextState = {
  speed: { value: 87, hasMultiValue: false },
};
const mockDispatch = jest.fn();

const mockCreateEventEmitter = jest.fn();
jest.mock('helpers/eventEmitterFactory', () => ({
  createEventEmitter: (...args) => mockCreateEventEmitter(...args),
}));
const mockEmit = jest.fn();

jest.mock('app/views/beambox/Right-Panels/contexts/LayerPanelContext', () => ({
  LayerPanelContext: React.createContext({ hasVector: false }),
}));

jest.mock('app/views/beambox/Right-Panels/contexts/ObjectPanelContext', () => ({
  ObjectPanelContext: React.createContext({ activeKey: null }),
}));

jest.mock('helpers/layer/check-vector', () => jest.fn());

describe('test SpeedBlock', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockCreateEventEmitter.mockReturnValueOnce({
      emit: mockEmit,
    });
  });

  it('should render correctly when unit is mm', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fbm1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          state: mockContextState as any,
          dispatch: mockDispatch,
          selectedLayers: mockSelectedLayers,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>
    );
    expect(mockStorageGet).toBeCalledTimes(1);
    expect(mockStorageGet).toHaveBeenLastCalledWith('default-units');
    expect(mockPrefRead).toBeCalledTimes(2);
    expect(mockPrefRead).toHaveBeenNthCalledWith(1, 'workarea');
    expect(mockPrefRead).toHaveBeenNthCalledWith(2, 'enable-low-speed');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when unit is inches', () => {
    mockStorageGet.mockReturnValueOnce('inches');
    mockPrefRead.mockReturnValueOnce('fbm1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          state: mockContextState as any,
          dispatch: mockDispatch,
          selectedLayers: mockSelectedLayers,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when type is panel-item', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fbm1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          state: mockContextState as any,
          dispatch: mockDispatch,
          selectedLayers: mockSelectedLayers,
        }}
      >
        <SpeedBlock type="panel-item" />
      </ConfigPanelContext.Provider>
    );
    expect(mockStorageGet).toBeCalledTimes(1);
    expect(mockStorageGet).toHaveBeenLastCalledWith('default-units');
    expect(mockPrefRead).toBeCalledTimes(2);
    expect(mockPrefRead).toHaveBeenNthCalledWith(1, 'workarea');
    expect(mockPrefRead).toHaveBeenNthCalledWith(2, 'enable-low-speed');
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when has vector warning', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fhex1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <LayerPanelContext.Provider value={{ hasVector: true } as any}>
        <ConfigPanelContext.Provider
          value={{
            state: mockContextState as any,
            dispatch: mockDispatch,
            selectedLayers: mockSelectedLayers,
          }}
        >
          <SpeedBlock />
        </ConfigPanelContext.Provider>
      </LayerPanelContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(mockPrefRead).toBeCalledTimes(3);
    expect(mockPrefRead).toHaveBeenNthCalledWith(1, 'workarea');
    expect(mockPrefRead).toHaveBeenNthCalledWith(2, 'enable-low-speed');
    expect(mockPrefRead).toHaveBeenNthCalledWith(3, 'vector_speed_contraint');
  });

  it('should render correctly when has low speed warning', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fhex1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const state = { ...mockContextState, speed: { value: 1 } };
    const { container } = render(
      <LayerPanelContext.Provider value={{ hasVector: true } as any}>
        <ConfigPanelContext.Provider
          value={{
            state: state as any,
            dispatch: mockDispatch,
            selectedLayers: mockSelectedLayers,
          }}
        >
          <SpeedBlock />
        </ConfigPanelContext.Provider>
      </LayerPanelContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(mockPrefRead).toBeCalledTimes(2);
    expect(mockPrefRead).toHaveBeenNthCalledWith(1, 'workarea');
    expect(mockPrefRead).toHaveBeenNthCalledWith(2, 'enable-low-speed');
  });

  test('onChange should work', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fbm1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          state: mockContextState as any,
          dispatch: mockDispatch,
          selectedLayers: mockSelectedLayers,
        }}
      >
        <SpeedBlock />
      </ConfigPanelContext.Provider>
    );
    expect(mockCreateEventEmitter).toBeCalledTimes(1);
    expect(mockCreateEventEmitter).toHaveBeenLastCalledWith('time-estimation-button');

    expect(mockDispatch).not.toBeCalled();
    expect(mockWriteData).not.toBeCalled();
    expect(mockEmit).not.toBeCalled();
    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: '88' } });
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      type: 'change',
      payload: { speed: 88, configName: 'CUSTOM_PRESET_CONSTANT' },
    });
    expect(mockWriteData).toBeCalledTimes(4);
    expect(mockWriteData).toHaveBeenNthCalledWith(1, 'layer1', 'speed', 88, true);
    expect(mockWriteData).toHaveBeenNthCalledWith(
      2,
      'layer1',
      'configName',
      'CUSTOM_PRESET_CONSTANT'
    );
    expect(mockWriteData).toHaveBeenNthCalledWith(3, 'layer2', 'speed', 88, true);
    expect(mockWriteData).toHaveBeenNthCalledWith(
      4,
      'layer2',
      'configName',
      'CUSTOM_PRESET_CONSTANT'
    );
    expect(mockEmit).toBeCalledTimes(1);
    expect(mockEmit).toHaveBeenLastCalledWith('SET_ESTIMATED_TIME', null);
  });

  test('onChange should work correctly when type is modal', () => {
    mockStorageGet.mockReturnValueOnce('mm');
    mockPrefRead.mockReturnValueOnce('fbm1').mockReturnValueOnce(true).mockReturnValueOnce(true);
    const { container } = render(
      <ConfigPanelContext.Provider
        value={{
          state: mockContextState as any,
          dispatch: mockDispatch,
          selectedLayers: mockSelectedLayers,
        }}
      >
        <SpeedBlock type="modal" />
      </ConfigPanelContext.Provider>
    );
    expect(mockCreateEventEmitter).toBeCalledTimes(1);
    expect(mockCreateEventEmitter).toHaveBeenLastCalledWith('time-estimation-button');

    expect(mockDispatch).not.toBeCalled();
    expect(mockWriteData).not.toBeCalled();
    expect(mockEmit).not.toBeCalled();
    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: '88' } });
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      type: 'change',
      payload: { speed: 88, configName: 'CUSTOM_PRESET_CONSTANT' },
    });
    expect(mockWriteData).not.toBeCalled();
  });
});
