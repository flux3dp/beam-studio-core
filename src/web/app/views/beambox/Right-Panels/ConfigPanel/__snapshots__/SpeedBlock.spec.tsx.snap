// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SpeedBlock should render correctly when has low speed warning 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      speed
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        speed
      </p>
      <p>
        min: 
        1
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        defaultValue: 
        1
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        displayMultiValue: 
      </p>
    </div>
    <input
      class="speed-for-vector"
      id="speed_value"
      max="300"
      min="1"
      step="0.1"
      type="range"
      value="1"
    />
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when has vector warning 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      speed
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        speed
      </p>
      <p>
        min: 
        1
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        defaultValue: 
        87
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        displayMultiValue: 
      </p>
    </div>
    <input
      class="speed-for-vector"
      id="speed_value"
      max="300"
      min="1"
      step="0.1"
      type="range"
      value="87"
    />
    <div
      class="warning"
    >
      <div
        class="warning-icon"
      >
        !
      </div>
      <div
        class="warning-text"
      >
        speed_contrain_warning
      </div>
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when type is panel-item 1`] = `
<div>
  <div>
    MockObjectPanelItem
    <div>
      id: 
      speed
    </div>
    <div>
      content: 
      <button
        class="adm-button adm-button-default adm-button-fill-outline adm-button-mini adm-button-shape-rounded number-item display-btn"
        type="button"
      >
        <span>
          87
        </span>
      </button>
    </div>
    <div>
      label: 
      speed
    </div>
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when unit is inches 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      speed
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        speed
      </p>
      <p>
        min: 
        1
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        unit: 
        in/s
      </p>
      <p>
        defaultValue: 
        87
      </p>
      <p>
        decimal: 
        2
      </p>
      <p>
        displayMultiValue: 
      </p>
    </div>
    <input
      class=""
      id="speed_value"
      max="300"
      min="1"
      step="0.1"
      type="range"
      value="87"
    />
  </div>
</div>
`;

exports[`test SpeedBlock should render correctly when unit is mm 1`] = `
<div>
  <div
    class="panel default"
  >
    <span
      class="title"
    >
      speed
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        speed
      </p>
      <p>
        min: 
        1
      </p>
      <p>
        max: 
        300
      </p>
      <p>
        unit: 
        mm/s
      </p>
      <p>
        defaultValue: 
        87
      </p>
      <p>
        decimal: 
        1
      </p>
      <p>
        displayMultiValue: 
      </p>
    </div>
    <input
      class=""
      id="speed_value"
      max="300"
      min="1"
      step="0.1"
      type="range"
      value="87"
    />
  </div>
</div>
`;
